render/respond.go

```diff
@@ -0,0 +1,111 @@
+package render
+
+import (
+	"bytes"
+	"encoding/json"
+	"encoding/xml"
+	"net/http"
+	"reflect"
+
+	"golang.org/x/net/context"
+)
+
+var Respond = DefaultRespond
+
+func DefaultRespond(ctx context.Context, w http.ResponseWriter, v interface{}) {
+	// Present the object.
+	if presenter, ok := ctx.Value(presenterCtxKey).(Presenter); ok {
+		v = presenter.Present(ctx, v)
+	} else {
+		v = DefaultPresenter.Present(ctx, v)
+	}
+
+	// Format data based on Content-Type.
+	switch contentType, _ := ctx.Value(contentTypeCtxKey).(ContentType); contentType {
+	case ContentTypeJSON:
+		JSON(ctx, w, v)
+	case ContentTypeXML:
+		XML(ctx, w, v)
+	default:
+		JSON(ctx, w, v)
+	}
+}
+
+func String(ctx context.Context, w http.ResponseWriter, v string) {
+	w.Header().Set("Content-Type", "text/plain; charset=utf-8")
+	status, _ := ctx.Value(statusCtxKey).(int)
+	if status == 0 {
+		status = 200
+	}
+	w.WriteHeader(status)
+	w.Write([]byte(v))
+}
+
+func HTML(ctx context.Context, w http.ResponseWriter, v string) {
+	w.Header().Set("Content-Type", "text/html; charset=utf-8")
+	status, _ := ctx.Value(statusCtxKey).(int)
+	if status == 0 {
+		status = 200
+	}
+	w.WriteHeader(status)
+	w.Write([]byte(v))
+}
+
+func JSON(ctx context.Context, w http.ResponseWriter, v interface{}) {
+	// Force to return empty JSON array [] instead of null in case of zero slice.
+	val := reflect.ValueOf(v)
+	if val.Kind() == reflect.Slice && val.IsNil() {
+		v = reflect.MakeSlice(val.Type(), 0, 0).Interface()
+	}
```

Comment:
We can have certain defaults that we feel are right for REST responder.

I think instead of responding with `nil` for array type, `[]` makes perfect sense and simplifies life of API consumers. But never mind -- if you guys overvote me, we can always put this outside of render pkg into custom `Render()` function that we're gonna use anyway. What do you think?